test/controllers/UploadControllerSpec.scala

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse, StatusCodes}
+import akka.stream.scaladsl.{Sink, Source}
+import akka.testkit.TestKit
+import controllers.auth.RequestWithOptionalEmpRef
+import helpers.ErsTestHelper
+import models.ERSFileProcessingException
+import models.upscan.{UploadId, UploadedSuccessfully, UpscanCsvFilesCallback, UpscanCsvFilesCallbackList}
+import org.mockito.ArgumentMatchers.{any, anyString, refEq}
+import org.mockito.Mockito.when
+import org.scalatestplus.play.guice.GuiceOneAppPerSuite
+import play.api.http.Status
+import play.api.i18n.{Messages, MessagesImpl}
+import play.api.inject.guice.GuiceApplicationBuilder
+import play.api.mvc.{AnyContent, DefaultMessagesControllerComponents, Request, Result}
+import play.api.test.FakeRequest
+import play.api.{Application, i18n}
+import services.{ProcessCsvService, ProcessODSService, StaxProcessor}
+import uk.gov.hmrc.http.{HeaderCarrier, UpstreamErrorResponse}
+import uk.gov.hmrc.play.test.UnitSpec
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, Future}
+
+class UploadControllerSpec extends TestKit(ActorSystem("UploadControllerTest")) with UnitSpec with ErsTestHelper with GuiceOneAppPerSuite{
```

Comment:
There is already an UploadControllerTest file, I'm not sure what the difference between that file and this?

Reply:
we didn't make it clear (will add comments to explain in the code):

UploadControllerTest is for the (unchanged) ods, whereas UploadControllerSpec is for the (changed) csv. when we were starting this work, we were hoping we would have enough time to clean up ods as well as csv, but we didn't want to start on it until we were fully finished with csv; in order to avoid having a massive mess of code in one test class that would need to ensure it tests the old, akka-less functionality for ods, and the new, akka-y functionality for csv, we moved the new tests to UploadControllerSpec